src/server/container_utils.h

```diff
@@ -0,0 +1,169 @@
+// Copyright 2022, Roman Gershman.  All rights reserved.
+// See LICENSE for licensing terms.
+//
+#pragma once
+
+extern "C" {
+#include "redis/intset.h"
+#include "redis/object.h"
+#include "redis/redis_aux.h"
+#include "redis/util.h"
+}
+
+#include <type_traits>
+
+#include "base/flags.h"
+#include "core/compact_object.h"
+#include "core/string_set.h"
+#include "server/common.h"
+
+ABSL_DECLARE_FLAG(bool, use_set2);
+
+namespace dfly {
+
+namespace container_utils {
+
+// StorageKind specifies how the container stores values.
+enum class StorageKind {
+  // Container stores all values as strings.
+  STRING,
+  // Container stores all values as integral values (int).
+  INTEGRAL,
+  // Container might store strings and integers.
+  MIXED
+};
+
+// StorageType is the respective C++ type of the container internals
+// to accommodate any of its values in the most optimal way.
+// STRING = std::string
+// INTEGRAL = int64_t
+// MIXED = std::variant<std::string, int64_t>
+template <StorageKind K>
+using StorageType =
+    std::conditional_t<K == StorageKind::MIXED, std::variant<std::string, int64_t>,
+                       std::conditional_t<K == StorageKind::STRING, std::string, int64_t>>;
+
+// IsContainer returns true if the iterator points to a container type.
+bool IsContainer(PrimeIterator it) {
+  unsigned type = it->second.ObjType();
+  return (type == OBJ_LIST || type == OBJ_SET || type == OBJ_ZSET);
+}
+
+StorageKind GetStorageKind(PrimeIterator it) {
+  unsigned type = it->second.ObjType();
+  unsigned encoding = it->second.Encoding();
+  switch (type) {
+    case OBJ_SET:
+      if (encoding == kEncodingIntSet)
+        return StorageKind::INTEGRAL;
+      else
+        return StorageKind::STRING;
+    default:
+      return StorageKind::MIXED;
+  };
+}
+
+// Create empty quicklistEntry
+quicklistEntry QLEntry() {
+  quicklistEntry res{.quicklist = NULL,
+                     .node = NULL,
+                     .zi = NULL,
+                     .value = NULL,
+                     .longval = 0,
+                     .sz = 0,
+                     .offset = 0};
+  return res;
+}
+
+// Iterate over all values and call func(val). Iteration stops as soon
+// as func return false.
+template <typename F> bool IterateList(PrimeIterator it, F&& func) {
+  quicklist* ql = static_cast<quicklist*>(it->second.RObjPtr());
+  quicklistIter* qiter = quicklistGetIteratorAtIdx(ql, AL_START_HEAD, 0);
+  quicklistEntry entry = QLEntry();
+  bool success = true;
+  while (success && quicklistNext(qiter, &entry)) {
```

Comment:
it looks that the code could be a bit simpler here.
1. instead of `while`, maybe use `do { } while (success)`.
2. There is no point in boolean AND since it only goes from true to false once. So replace `&=` with `=`.